fdbserver/DataDistribution.actor.cpp

```diff
@@ -148,10 +150,29 @@ ACTOR Future<Void> updateServerMetrics( TCServerInfo *server ) {
 			}
 		}
 	}
+
+	if(cx.getPtr()) {
+		Version versionNow = wait(runRYWTransaction(cx, [](Reference<ReadYourWritesTransaction> tr) {
+								tr->setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+								return tr->getReadVersion(); }));
+		Version versionDiff = versionNow - server->serverMetrics.get().version;
```

Comment:
In a fearless cluster, SS in the remote DC can have a large `versionDiff` because its tLog in remote DC cannot pull data quickly enough from the primary DC. 

If WAN between two DCs have long latency or low throughput, all SSes in remote will have large `versionDiff` and be marked as failed.

I don't think that is a desired action. 

If we are solving the issue https://github.com/apple/foundationdb/issues/1510, we probably need to calculate the version diff. from the tLog's version?

Reply:
The version difference allowed is a knob and the default value is pretty large, i.e, 250 seconds. It can be adjusted according to the expected latency or throughput. Under normal operations 250 seconds is big enough to allow the current calculation. 

The worst case version difference between TLogs in a DC is 5 seconds. In the scenarios like where the whole DC is lagging, all the storage servers on the remote will be marked as lagging and hence data movement will not be triggered. @etschannen, do you have an opinion on this?

Reply:
In the scenario that WAN across DC is down for > 250 seconds, all SS in the remote DC will be marked as lagged and later failed in the current commit. This means the remote DC will have lots of SS removed.

Later, when WAN is fixed, we need to not only let tLog catch up the lag, but also recreate those SSes in the remote DC. That's  my concern here.

Reply:
If i understand correctly, if the remote DC is down for a while and comes back all the SS in the remote will be marked as Undesired. But since there is no immediate desirable team, i don't think DD will start moving data. But as and when the storage servers catch up, there is a chance that some teams catchup quickly than others and some data movement can be queued. But as the other teams also catchup, then the queued data movement should be cancelled. I agree that there is a possibility of some extra data movement although i think it should be minimal. I will also bring this up with @etschannen to get his input.

Reply:
Your understanding is correct.

My intuition is that the data movement could be a lot in corner cases.
DD can queue a lot of RelocateShard requests before actually move it. It is relatively fast to create RelocateShard requests. So as long as there is some time gap between remote SSes' catching up, there could have already queued a lot of RelocateShard requests.

Another intuition is that relying on "there is no desired team in the remote" to prevent the problem (e..g, extra excessive data movement and losing lots of SSes) seems risky to me.

Reply:
@xumengpanda  I brought up this concern you raised with @etschannen and we are planning to go with your suggestion of fetching version difference. There are also going to be other changes that will limit the impact of this algorithm when a large number of storage servers are lagging. Thanks.

Reply:
Great! looking forward to the change!